README.md

```diff
@@ -198,289 +189,494 @@ e.increment();  // logs 1
+      for (const {name, eventName} of handlers) {
+        this.addEventListener(eventName, this[name].bind(this));
+      }
+    }
+  }
 }
-C.prototype.method = f(C.prototype.method);
 ```
 
-`@wrap` can also be used on a class to wrap the entire class.
+#### `@bound` with a mixin
+
+`@bound` could be used with a mixin superclass as follows:
 
 ```js
-@wrap(f)
-class C { }
+class C extends WithBoundMethod(Object) {
+  #x = 1;
+  @bound method() { return this.#x; }
+}
+
+let c = new C;
+let m = c.method;
+m();  // 1, not TypeError
 ```
 
-is roughly equivalent to:
+This decorator could be defined as:
 
 ```js
-class C { }
-C = f(C);
+const boundName = Symbol("boundName");
+function bound(method, context) {
+  context.metadata = {[boundName]: true};
+  return method;
+}
+let boundMap = new MetadataLookupCache(boundName);
+
+function WithBoundMethods(superclass) {
+  return class C extends superclass {
+    constructor(...args) {
+      super(...args);
+      let names = boundMap.get(new.target, C);
+      for (const {name} of names) {
+        this[name] = this[name].bind(this);
+      }
+    }
+  }
+}
 ```
 
-Details:
-- `@wrap` may be used on private methods as well as public ones, static as well as instance.
-- The function is only passed the method, and no other context.
-- The return value is used to replace the method or accessor.
-- `@wrap` may be used on getters or setters, and applies to these individually.
-- `@wrap` may not be used on field declarations, as there's no clear meaning.
-- When `@wrap` is used on a class, if there is a use of `C` in a method or field initializer inside the class, it will refer to the original, unwrapped `C. See [#211](https://github.com/tc39/proposal-decorators/issues/211) for details.
+Note the common use of `MetadataLookupCache` across both examples; this proposal or a follow-on one should consider adding a standard library for adding metadata for this purpose.
+
+### Option B: The `init` contextual keyword for methods
 
-### `@register`
+If it's not acceptable to require a superclass/mixin for cases requiring initialization action, an The `init` keyword in a method declaration changes a method into an "init method". This keyword allows decorators to add initialization actions, run when the constructor executes.
 
-The `@register` decorator schedules a callback to run after the class is created.
+#### `@on` with `init`
+
+Usage:
 
 ```js
-class C {
-  @register(f) method() { }
+class MyElement extends HTMLElement {
+  @on('click') init clickHandler() { }
 }
 ```
 
-is roughly equivalent to
+An "init method" (method declared with `init`) is called similarly to a method decorator, but it is expected to return a pair `{method, initialize}`, where `initialize` is called with the `this` value being the new instance, taking no arguments and returning nothing. 
 
 ```js
-class C {
-  method() { }
+function on(eventName) {
+  return (method, context) => {
+    assert(context.kind === "init-method");
```

Comment:
Just got another question. If I apply `@on init` to a property what would I get in `context.kind`? `init-field`?

```js
class Foo extends HTMLElement {
  @on('click') init bar = 1;
}
```

Reply:
SyntaxError

Reply:
Oh I see. We already have an `initialize` property for `field` type, so using `init` is unnecessary here. Got it.